### PR TITLE
Wire creator contracts

### PR DESCRIPTION
**Pull request by vince0656**
https://github.com/knownorigin/known-origin-discovery/issues/1067

### CODE REVIEW

File: abis/kodav4/BatchCreatorContract.json [hunk 1]
```diff
@@ -0,0 +1,2275 @@
+[
```
Comment (jamesmorgan) [inline abis/kodav4/BatchCreatorContract.json:1]: Wondering if these names should be specific to the actual impl?

File: abis/kodav4/CreatorFactory.json [hunk 2]
```diff
@@ -0,0 +1,784 @@
+[
```
Comment (jamesmorgan) [inline abis/kodav4/CreatorFactory.json:1]: Same for this in terms of file naming?

File: abis/kodav4/SelfSovereignFundsHandler.json [hunk 3]
```diff
@@ -0,0 +1,150 @@
+[
```
Comment (jamesmorgan) [inline abis/kodav4/SelfSovereignFundsHandler.json:1]: Do we need this vs using the existing funds handlers?

File: schema.graphql [hunk 4]
```diff
@@ -533,7 +543,59 @@ type GatedSale @entity {
...
+
+    # Current Settings
+    paused: Boolean!
+    owner: Bytes!
+    secondaryRoyaltyPercentage: BigInt!
+    minter: Bytes!
+
+    # Funds handler
+    defaultFundsHandler: Bytes!
+    defaultFundsRecipients: [Bytes!]!
```
Comment (jamesmorgan) [inline schema.graphql:565]: Should this be a `Collective`?

File: src/mappings/creator-contracts/creator-contract.ts [hunk 5]
```diff
@@ -0,0 +1,500 @@
...
+import {addEditionToArtist, recordArtistValue} from "../../services/Artist.service";
+import {loadOrCreateListedToken} from "../../services/ListedToken.service";
+import {toEther} from "../../utils/utils";
+import {ONE, ONE_ETH, ZERO, ZERO_ADDRESS, ZERO_BIG_DECIMAL} from "../../utils/constants";
+
+export function handleEditionSalesDisabledUpdated(event: EditionSalesDisabledUpdated): void {
+    let contractEntity = CreatorContract.load(event.address.toHexString())
+    let editionEntity = loadOrCreateV4Edition(event.params._editionId, event.block, event.address, contractEntity.isHidden);
+    editionEntity.active = event.params._disabled;
+    editionEntity.save()
```
Comment (jamesmorgan) [inline src/mappings/creator-contracts/creator-contract.ts:68]: This needs an activity event I think

File: src/mappings/creator-contracts/creator-contract.ts [hunk 6]
```diff
@@ -0,0 +1,500 @@
...
+    let editionEntity = loadOrCreateV4Edition(event.params._editionId, event.block, event.address, contractEntity.isHidden);
+
+    let creatorContractInstance = ERC721CreatorContract.bind(event.address)
+    populateEditionMetadata(
+        editionEntity,
+        editionEntity.id,
+        creatorContractInstance.editionURI(event.params._editionId)
+    )
+
+    editionEntity.save()
```
Comment (jamesmorgan) [inline src/mappings/creator-contracts/creator-contract.ts:82]: Activity event

File: src/mappings/creator-contracts/creator-contract.ts [hunk 7]
```diff
@@ -0,0 +1,500 @@
...
+}
+
+export function handlePaused(event: Paused): void {
+    let entity = CreatorContract.load(event.address.toHexString());
+    entity.paused = true;
+    entity.save();
```
Comment (jamesmorgan) [inline src/mappings/creator-contracts/creator-contract.ts:88]: Activity event

File: src/mappings/creator-contracts/creator-contract.ts [hunk 8]
```diff
@@ -0,0 +1,500 @@
...
+}
+
+export function handleUnpaused(event: Unpaused): void {
+    let entity = CreatorContract.load(event.address.toHexString());
+    entity.paused = false;
+    entity.save();
```
Comment (jamesmorgan) [inline src/mappings/creator-contracts/creator-contract.ts:94]: Activity event

File: src/mappings/creator-contracts/creator-contract.ts [hunk 9]
```diff
@@ -0,0 +1,500 @@
...
+        event.block,
+        event.address,
+        contractEntity.isHidden
+    )
+
+    edition.startDate = event.params._startDate
+    edition.priceInWei = event.params._price
+    edition.salesType = SaleTypes.BUY_NOW
+
+    edition.save()
```
Comment (jamesmorgan) [inline src/mappings/creator-contracts/creator-contract.ts:195]: Activity event

File: src/mappings/creator-contracts/creator-contract.ts [hunk 10]
```diff
@@ -0,0 +1,500 @@
...
+        event.block,
+        event.address,
+        contractEntity.isHidden
+    )
+
+    edition.startDate = ZERO
+    edition.priceInWei = ZERO
+    edition.salesType = SaleTypes.OFFERS_ONLY // Revert back to the default state during de-listing
+
+    edition.save()
```
Comment (jamesmorgan) [inline src/mappings/creator-contracts/creator-contract.ts:211]: Activity event

File: src/mappings/creator-contracts/creator-contract.ts [hunk 11]
```diff
@@ -0,0 +1,500 @@
...
+    recordArtistValue(creator, tokenEntity.id, event.params._price, artistShareOfETHInWei, true);
+
+    // Update token events
+    tokenEventFactory.createTokenPrimaryPurchaseEvent(event, tokenEntity.id, event.params._buyer, event.params._price);
+}
+
+export function handleOwnershipTransferred(event: OwnershipTransferred): void {
+    let creatorContractEntity = CreatorContract.load(event.address.toHexString())
+    creatorContractEntity.owner = event.params.newOwner
+    creatorContractEntity.save()
```
Comment (jamesmorgan) [inline src/mappings/creator-contracts/creator-contract.ts:294]: Activity event

File: src/mappings/creator-contracts/creator-contract.ts [hunk 12]
```diff
@@ -0,0 +1,500 @@
...
+}
+
+export function handleSecondaryRoyaltyUpdated(event: DefaultRoyaltyPercentageUpdated): void {
+    let creatorContractEntity = CreatorContract.load(event.address.toHexString())
+    creatorContractEntity.secondaryRoyaltyPercentage = event.params._percentage
+    creatorContractEntity.save()
```
Comment (jamesmorgan) [inline src/mappings/creator-contracts/creator-contract.ts:300]: Activity event

File: src/mappings/creator-contracts/creator-contract.ts [hunk 13]
```diff
@@ -0,0 +1,500 @@
...
+}
+
+export function handleSecondaryEditionRoyaltyUpdated(event: EditionRoyaltyPercentageUpdated): void {
+    let entityId = event.params._editionId.toString() + "-" + event.address.toHexString()
+    let entity = Edition.load(entityId)
+    if (entity != null) {
+        entity.secondaryRoyaltyV4EditionOverride = event.params._percentage
+        entity.save()
+    }
```
Comment (jamesmorgan) [inline src/mappings/creator-contracts/creator-contract.ts:309]: Activity event

File: src/mappings/creator-contracts/creator-contract.ts [hunk 14]
```diff
@@ -0,0 +1,500 @@
...
+
+    collective.recipients = creatorContractEntity.defaultFundsRecipients
+    collective.splits = creatorContractEntity.defaultFundsShares
+
+    collective.save()
+
+    edition.collective = editionFundsHandler
+    edition.save()
+}
```
Comment (jamesmorgan) [inline src/mappings/creator-contracts/creator-contract.ts:367]: Activity event

File: src/mappings/creator-contracts/creator-contract.ts [hunk 15]
```diff
@@ -0,0 +1,500 @@
...
+    // update token info
+    let tokenEntity = loadOrCreateV4Token(event.params._tokenId, event.address, edition, event.block);
+    tokenEntity.isListed = true;
+    tokenEntity.salesType = SaleTypes.BUY_NOW
+    tokenEntity.listPrice = toEther(event.params._price)
+    tokenEntity.lister = creatorContractInstance.ownerOf(event.params._tokenId).toHexString()
+    tokenEntity.listingTimestamp = event.block.timestamp
+    tokenEntity.openOffer = null
+    tokenEntity.currentTopBidder = null
+    tokenEntity.save()
```
Comment (jamesmorgan) [inline src/mappings/creator-contracts/creator-contract.ts:398]: Activity event

File: src/mappings/creator-contracts/creator-contract.ts [hunk 16]
```diff
@@ -0,0 +1,500 @@
...
+
+    let tokenEntity = loadOrCreateV4Token(event.params._tokenId, event.address, edition, event.block);
+    tokenEntity.isListed = false;
+    tokenEntity.salesType = SaleTypes.OFFERS_ONLY;
+    tokenEntity.listPrice = ZERO_BIG_DECIMAL;
+    tokenEntity.lister = null;
+    tokenEntity.listingTimestamp = ZERO;
+    tokenEntity.openOffer = null;
+    tokenEntity.currentTopBidder = null;
+    tokenEntity.save();
```
Comment (jamesmorgan) [inline src/mappings/creator-contracts/creator-contract.ts:421]: Activity event

File: src/mappings/creator-contracts/creator-contract.ts [hunk 17]
```diff
@@ -0,0 +1,500 @@
...
+    );
+
+    let tokenEntity = loadOrCreateV4Token(event.params._tokenId, event.address, edition, event.block);
+    tokenEntity.listPrice = toEther(event.params._price);
+    tokenEntity.save();
+
+    let entityId = event.params._tokenId.toString() + '-' + event.address.toHexString();
+    let listedEntity = loadOrCreateListedToken(entityId, edition);
+    listedEntity.listPrice = toEther(event.params._price);
+    listedEntity.save();
```
Comment (jamesmorgan) [inline src/mappings/creator-contracts/creator-contract.ts:440]: Activity event

File: src/mappings/creator-contracts/creator-contract.ts [hunk 18]
```diff
@@ -0,0 +1,500 @@
...
+
+    // activity event
+    activityEventService.recordSecondarySale(
+        event,
+        tokenEntity,
+        edition,
+        event.params._price,
+        event.params._recipient,
+        event.params._currentOwner
+    );
```
Comment (jamesmorgan) [inline src/mappings/creator-contracts/creator-contract.ts:499]: Activity event

File: src/mappings/creator-contracts/creator-contract.ts [hunk 19]
```diff
@@ -0,0 +1,500 @@
...
+}
```
Comment (jamesmorgan) [inline src/mappings/creator-contracts/creator-contract.ts:500]: All looks good but I think we need activity events for all events in which we will need to update marketplace state. We in essence scrap events and then use this to build a model of the editions/tokens. So any changes to these need an event I think

File: src/mappings/creator-contracts/factory.ts [hunk 20]
```diff
@@ -0,0 +1,117 @@
...
+    creatorContractEntity.defaultFundsShares = defaultFundsShares
+
+    creatorContractEntity.save()
+
+    // Update the artist
+    let artistEntity = loadOrCreateArtist(event.params.artist)
+    let creatorContracts = artistEntity.creatorContracts
+    creatorContracts.push(event.params.selfSovereignNFT.toHexString())
+    artistEntity.creatorContracts = creatorContracts
+    artistEntity.save()
```
Comment (jamesmorgan) [inline src/mappings/creator-contracts/factory.ts:110]: Activity event

File: src/mappings/creator-contracts/factory.ts [hunk 21]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+export function handleCreatorContractBanned(event: CreatorContractBanned): void {
+    let creatorContractEntity = CreatorContract.load(event.params._contract.toHexString())
+    creatorContractEntity.isHidden = true
+    creatorContractEntity.save()
```
Comment (jamesmorgan) [inline src/mappings/creator-contracts/factory.ts:116]: Activity event

**Review by jamesmorgan — CHANGES_REQUESTED** [review]

File: abis/kodav4/ClaimableFundsReceiverV1.json [hunk 22]
```diff
@@ -0,0 +1,150 @@
+[
```
Comment (jamesmorgan) [inline abis/kodav4/ClaimableFundsReceiverV1.json:1]: Do we need this - I think the assumption should be that funds handlers will be the V3 deployed flow for now so I dont think we need this ABI or logic in here atm

File: src/services/ActivityEvent.service.ts [hunk 23]
```diff
@@ -487,3 +497,110 @@ export function recordGatedSaleResumed(rawEvent: ethereum.Event, sale: GatedSale
 
     event.save()
 }
+
+function createCreatorContractEventId(address: string, id: string, event: ethereum.Event): string {
+    return "CreatorContract"
```
Comment (jamesmorgan) [inline src/services/ActivityEvent.service.ts:502]: Lets make this a constant for `"CreatorContract"`

File: src/services/ActivityEvent.service.ts [hunk 24]
```diff
@@ -487,3 +497,110 @@ export function recordGatedSaleResumed(rawEvent: ethereum.Event, sale: GatedSale
...
+        .concat(event.transaction.hash.toHexString())
+        .concat("-")
+        .concat(event.logIndex.toString());
+}
+
+function createdCreatorContractEvent(ID: string, type: string, rawEvent: ethereum.Event, edition: Edition | null): ActivityEvent {
+    let event: ActivityEvent = new ActivityEvent(ID);
+
+    event.version = edition ? edition.version : BigInt.fromString("4")
+    event.type = edition ? TYPE_EDITION : "CreatorContract" // For V4, we're either dealing with an edition or something at the global contract level
```
Comment (jamesmorgan) [inline src/services/ActivityEvent.service.ts:517]: Constant

File: schema.graphql [hunk 25]
```diff
@@ -242,6 +242,8 @@ type Token @entity {
 
     # Complete attached token
     listing: ListedToken
+
+    creatorContract: Bytes # Optionally capture creator contract address if it is version 4
```
Comment (jamesmorgan) [inline schema.graphql:246]: Could these be object mappings? `CreatorContract` and not `Bytes`?

File: schema.graphql [hunk 26]
```diff
@@ -292,6 +294,8 @@ type ListedToken @entity {
 
     reserveAuctionCanEmergencyExit: Boolean!
     # ----- End of reserve auction fields - #
+
+    creatorContract: Bytes # Optionally capture creator contract address if this is version 4
```
Comment (jamesmorgan) [inline schema.graphql:298]: Could these be object mappings? `CreatorContract` and not `Bytes`?

File: schema.graphql [hunk 27]
```diff
@@ -332,10 +336,11 @@ type Collective @entity {
 # A token/editions and its meta data
 type Edition @entity {
     id: ID!
-    version: BigInt! # KODA 2 or 3
+    version: BigInt! # KODA 2 or 3 or 4
     editionNmber: BigInt! # edition no. as Bigint
     editionType: BigInt!
     editionData: Bytes # bytes32 - some data about the edition
+    creatorContract: Bytes # address of creator contract that minted the edition if version 4
```
Comment (jamesmorgan) [inline schema.graphql:343]: Could these be object mappings? `CreatorContract` and not `Bytes`?

File: schema.graphql [hunk 28]
```diff
@@ -361,12 +366,15 @@ type Edition @entity {
     totalBurnt: BigInt!          # number of items send to 0xDead or 0x0 (only applicable for V3)
     originalEditionSize: BigInt! # original edition size of the edition at mint - this field should NEVER change once set
 
+    secondaryRoyaltyV4EditionOverride: BigInt # At the edition level, the EIP2981 percentage requested by the creator(s)
```
Comment (jamesmorgan) [inline schema.graphql:369]: Is this not the same as `artistCommission` ?

File: schema.graphql [hunk 29]
```diff
@@ -533,7 +541,59 @@ type GatedSale @entity {
...
 }
 
-#########################
+# Any KODA V4 Contract Deployed from the V4 factory
+type CreatorContract @entity {
+    id: ID!
+
+    # Deployment information
+    deploymentBlockNumber: BigInt!
+    deploymentTimestamp: BigInt!
```
Comment (jamesmorgan) [inline schema.graphql:550]: Common deployment properties used on other entities 

```
    # Common event details
    timestamp: BigInt!
    transactionHash: Bytes!
    transactionIndex: BigInt!
    logIndex: BigInt!
    eventAddress: Bytes!
    eventTxFrom: Bytes!
    eventTxTo: Bytes
    blockNumber: BigInt!
```

Port over the ones which make sense and use same names if possible

File: src/mappings/creator-contracts/creator-contract.ts [hunk 30]
```diff
@@ -0,0 +1,605 @@
+import {Address, BigDecimal, BigInt, Bytes, log, store} from "@graphprotocol/graph-ts";
```
Comment (jamesmorgan) [inline src/mappings/creator-contracts/creator-contract.ts:1]: Wondering if this should within the `v4` folder and not `creator-contracts` for consistancy?

File: src/mappings/creator-contracts/creator-contract.ts [hunk 31]
```diff
@@ -0,0 +1,605 @@
...
+        event,
+        tokenEntity,
+        edition,
+        event.params._price,
+        creatorContractInstance.ownerOf(event.params._tokenId)
+    );
+}
+
+export function handleBuyNowTokenDeListed(event: BuyNowTokenDeListed): void {
+    let entityId = event.params._tokenId.toString() + '-' + event.address.toHexString();
```
Comment (jamesmorgan) [inline src/mappings/creator-contracts/creator-contract.ts:490]: I think we should have a common utils file/method which takes an ID and address and builds the composite key - prevents any issues arising form doing it in the wrong way some like this `buildCCKey(id, address)`

File: src/mappings/creator-contracts/creator-contract.ts [hunk 32]
```diff
@@ -0,0 +1,605 @@
...
+        event.block,
+        event.address,
+        contractEntity.isHidden
+    );
+
+    let tokenEntity = loadOrCreateV4Token(event.params._tokenId, event.address, edition, event.block);
+    tokenEntity.listPrice = toEther(event.params._price);
+    tokenEntity.save();
+
+    let entityId = event.params._tokenId.toString() + '-' + event.address.toHexString();
```
Comment (jamesmorgan) [inline src/mappings/creator-contracts/creator-contract.ts:533]: Common key

File: src/mappings/creator-contracts/creator-contract.ts [hunk 33]
```diff
@@ -0,0 +1,605 @@
...
+        event,
+        tokenEntity,
+        edition,
+        event.params._price,
+        creatorContractInstance.ownerOf(event.params._tokenId),
+    );
+}
+
+export function handleBuyNowTokenPurchased(event: BuyNowTokenPurchased): void {
+    let entityId = event.params._tokenId.toString() + '-' + event.address.toHexString();
```
Comment (jamesmorgan) [inline src/mappings/creator-contracts/creator-contract.ts:549]: Common key

**Review by jamesmorgan — CHANGES_REQUESTED** [review]
Looking good - some minor points

File: src/services/Token.service.ts [hunk 34]
```diff
@@ -155,6 +155,32 @@ export function loadOrCreateV3Token(tokenId: BigInt, contract: KnownOriginV3, bl
...
+    tokenEntity.tokenURI = edition.tokenURI
+    tokenEntity.metadata = edition.metadata
+    tokenEntity.birthTimestamp = block.timestamp
+    tokenEntity.lastTransferTimestamp = block.timestamp
+    tokenEntity.editionActive = edition.active
+    return tokenEntity
+}
+
+export function loadOrCreateV4Token(tokenId: BigInt, contractAddress: Address, edition: Edition, block: ethereum.Block): Token {
+    let entityId = tokenId.toString() + "-" + contractAddress.toHexString()
```
Comment (jamesmorgan) [inline src/services/Token.service.ts:172]: For some reason I would have assumed the other way round would be better `address-tokenId` as this is where the natural grouping lies - should be change?

File: abis/kodav4/ClaimableFundsReceiverV1.json (hunk 22 above)
Comment (jamesmorgan) [reply abis/kodav4/ClaimableFundsReceiverV1.json:1]: Keep it - I was more thinking that the ABI is wrong not the logic i.e. should it be the V3 claimable funds receiver?

File: src/services/Token.service.ts (hunk 34 above)
Comment (jamesmorgan) [reply src/services/Token.service.ts]: I think liam already did this yesterday but worth pulling and checking

**Review by liamcf44 — APPROVED** [review]

**Review by jamesmorgan — CHANGES_REQUESTED** [review]
LGTM - minor comment on burns

**Review by jamesmorgan — APPROVED** [review]

**Pull request merged**